warpsql-plugins/extern/gearmand/man/gearmand.8
.\" Man page generated from reStructuredText.
.
.TH "GEARMAND" "8" "February 11, 2014" "1.1.12" "Gearmand"
.SH NAME
gearmand \- Gearmand Documentation, http://gearman.info/
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBGeneral options\fP
.INDENT 0.0
.TP
.B \-b [ \-\-backlog ] arg (=32)
Number of backlog connections for listen.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-check\-args
Check command line and configuration file argments and then exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-d [ \-\-daemon ]
Daemon, detach and run in the background.
.UNINDENT
.INDENT 0.0
.TP
.B \-f [ \-\-file\-descriptors ] arg
Number of file descriptors to allow for the process (total connections will be slightly less). Default is max allowed for user.
.UNINDENT
.INDENT 0.0
.TP
.B \-h [ \-\-help ]
Print this help menu.
.UNINDENT
.INDENT 0.0
.TP
.B \-j [ \-\-job\-retries ] arg (=0)
Number of attempts to run the job before the job server removes it. This is helpful to ensure a bad job does not crash all available workers. Default is no limit.
.UNINDENT
.INDENT 0.0
.TP
.B \-l [ \-\-log\-file ] arg
Log file to write errors and information to.  Turning this option on also forces the first verbose level to be enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-L [ \-\-listen ] arg
Address the server should listen on. Default is INADDR_ANY.
.UNINDENT
.INDENT 0.0
.TP
.B \-p [ \-\-port ] arg (=4730)
Port the server should listen on.
.UNINDENT
.INDENT 0.0
.TP
.B \-P [ \-\-pid\-file ] arg
File to write process ID out to.
.UNINDENT
.INDENT 0.0
.TP
.B \-r [ \-\-protocol ] arg
Load protocol module.
.UNINDENT
.INDENT 0.0
.TP
.B \-R [ \-\-round\-robin ]
Assign work in round\-robin order per worker connection. The default is to assign work in the order of functions added by the worker.
.UNINDENT
.INDENT 0.0
.TP
.B \-q [ \-\-queue\-type ] arg
Persistent queue type to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-t [ \-\-threads ] arg (=4)
Number of I/O threads to use. Default=4.
.UNINDENT
.INDENT 0.0
.TP
.B \-u [ \-\-user ] arg
Switch to given user after startup.
.UNINDENT
.INDENT 0.0
.TP
.B \-v [ \-\-verbose ] arg (=v)
Increase verbosity level by one.
.UNINDENT
.INDENT 0.0
.TP
.B \-V [ \-\-version ]
Display the version of gearmand and exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-w [ \-\-worker\-wakeup ] arg (=0)
Number of workers to wakeup for each job received. The default is to wakeup all available workers.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-keepalive
Enable keepalive on sockets.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-keepalive\-idle arg (=\-1)
If keepalive is enabled, set the value for TCP_KEEPIDLE for systems that support it. A value of \-1 means that either the system does not support it or an error occurred when trying to retrieve the default value.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-keepalive\-interval arg (=\-1)
If keepalive is enabled, set the value for TCP_KEEPINTVL for systems that support it. A value of \-1 means that either the system does not support it or an error occurred when trying to retrieve the default value.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-keepalive\-count arg (=\-1)
If keepalive is enabled, set the value for TCP_KEEPCNT for systems that support it. A value of \-1 means that either the system does not support it or an error occurred when trying to retrieve the default value.
.UNINDENT
.sp
\fBHTTP:\fP
.INDENT 0.0
.TP
.B \-\-http\-port arg (=8080)
Port to listen on.
.UNINDENT
.sp
\fBsqlite\fP
.INDENT 0.0
.TP
.B \-\-libsqlite3\-db arg
Database file to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libsqlite3\-table arg (=gearman_queue)
Table to use.
.UNINDENT
.sp
\fBMemcached(libmemcached)\fP
.INDENT 0.0
.TP
.B \-\-libmemcached\-servers arg
List of Memcached servers to use.
.UNINDENT
.sp
\fBDrizzle/MySQL(libdrizzle)\fP
.INDENT 0.0
.TP
.B \-\-libdrizzle\-host arg
Host of server.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-port arg
Port of server. (by default Drizzle)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-uds arg
Unix domain socket for server.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-user arg
User name for authentication.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-password arg
Password for authentication.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-db arg
Schema/Database to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-table arg
Table to use.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libdrizzle\-mysql arg
Use MySQL protocol.
.UNINDENT
.sp
\fBPostgres\fP
.INDENT 0.0
.TP
.B \-\-libpq\-conninfo arg
PostgreSQL connection information string.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libpq\-table arg (=queue)
Table to use.
.UNINDENT
.sp
\fBtokyocabinet\fP
.INDENT 0.0
.TP
.B \-\-libtokyocabinet\-file arg
File name of the database. [see: man tcadb, tcadbopen() for name guidelines]
.UNINDENT
.INDENT 0.0
.TP
.B \-\-libtokyocabinet\-optimize
Optimize database on open. [default=true]
.UNINDENT
.SH DESCRIPTION
.sp
Gearman provides a generic application framework to farm out work to other machines or processes that are better suited to do the work. It allows you to do work in parallel, to load balance processing, and to call functions between languages. It can be used in a variety of applications, from high\-availability web sites to the transport of database replication events. In other words, it is the nervous system for how distributed processing communicates. A few strong points about Gearman:
.INDENT 0.0
.IP \(bu 2
Open Source \- It\(aqs free! (in both meanings of the word) Gearman has an active open source community that is easy to get involved with if you need help or want to contribute.
.IP \(bu 2
Multi\-language \- There are interfaces for a number of languages, and this list is growing. You also have the option to write heterogeneous applications with clients submitting work in one language and workers performing that work in another.
.IP \(bu 2
Flexible \- You are not tied to any specific design pattern. You can quickly put together distributed applications using any model you choose, one of those options being Map/Reduce.
.IP \(bu 2
Fast \- Gearman has a simple protocol and interface with a new optimized server in C to minimize your application overhead.
.IP \(bu 2
Embeddable \- Since Gearman is fast and lightweight, it is great for applications of all sizes. It is also easy to introduce into existing applications with minimal overhead.
.IP \(bu 2
No single point of failure \- Gearman can not only help scale systems, but can do it in a fault tolerant way.
.UNINDENT
.SS Thread Model
.sp
The \-t option to gearmand allows you to specify multiple I/O threads, this is enabled by default. There are currently three types of threads in the job server:
.sp
Listening and management thread \- only one
I/O thread \- can have many
Processing thread \- only one
.sp
When no \-t option is given or \-t 0 is given, all of three thread types happen within a single thread. When \-t 1 is given, there is a thread for listening/management and a thread for I/O and processing. When \-t 2 is given, there is a thread for each type of thread above. For all \-t option values above 2, more I/O threads are created.
.sp
The listening and management thread is mainly responsible for accepting new connections and assigning those connections to an I/O thread (if there are many). It also coordinates startup and shutdown within the server. This thread will have an instance of libevent for managing socket events and signals on an internal pipe. This pipe is used to wakeup the thread or to coordinate shutdown.
.sp
The I/O thread is responsible for doing the read and write system calls on the sockets and initial packet parsing. Once the packet has been parsed it it put into an asynchronous queue for the processing thread (each thread has it\(aqs own queue so there is very little contention). Each I/O thread has it\(aqs own instance of libevent for managing socket events and signals on an internal pipe like the listening thread.
.sp
The processing thread should have no system calls within it (except for the occasional brk() for more memory), and manages the various lists and hash tables used for tracking unique keys, job handles, functions, and job queues. All packets that need to be sent back to connections are put into an asynchronous queue for the I/O thread. The I/O thread will pick these up and send them back over the connected socket. All packets flow through the processing thread since it contains the information needed to process the packets. This is due to the complex nature of the various lists and hash tables. If multiple threads were modifying them the locking overhead would most likely cause worse performance than having it in a single thread (and would also complicate the code). In the future more work may be pushed to the I/O threads, and the processing thread can retain minimal functionality to manage those tables and lists. So far this has not been a significant bottleneck, a 16 core Intel machine is able to process upwards of 50k jobs per second.
.sp
For thread safety to work when UUID are generated, you must be running the uuidd daemon.
.SS Persistent Queues
.sp
Inside the Gearman job server, all job queues are stored in memory. This means if a server restarts or crashes with pending jobs, they will be lost and are never run by a worker. Persistent queues were added to allow background jobs to be stored in an external durable queue so they may live between server restarts and crashes. The persistent queue is only enabled for background jobs because foreground jobs have an attached client. If a job server goes away, the client can detect this and restart the foreground job somewhere else (or report an error back to the original caller). Background jobs on the other hand have no attached client and are simply expected to be run when submitted.
.sp
The persistent queue works by calling a module callback function right before putting a new job in the internal queue for pending jobs to be run. This allows the module to store the job about to be run in some persistent way so that it can later be replayed during a restart. Once it is stored through the module, the job is put onto the active runnable queue, waking up available workers if needed. Once the job has been successfully completed by a worker, another module callback function is called to notify the module the job is done and can be removed. If a job server crashes or is restarted between these two calls for a job, the jobs are reloaded during the next job server start. When the job server starts up, it will call a replay callback function in the module to provide a list of all jobs that were not complete. This is used to populate the internal memory queue of jobs to be run. Once this replay is complete, the job server finishes its initialization and the jobs are now runnable once workers connect (the queue should be in the same state as when it crashed). These jobs are removed from the persistent queue when completed as normal. NOTE: Deleting jobs from the persistent queue storage will not remove them from the in\-memory queue while the server is running.
.sp
The queues are implemented using a modular interface so it is easy to add new data stores for the persistent queue.
.sp
A persistent queue module is enabled by passing the \-q or –queue\-type option to gearmand. Run gearmand –help to see which queue modules are supported on your system. If you are missing options for one you would like to use, you will need to install any dependencies and then recompile the gearmand package.
.SS Extended Protocols
.sp
The protocol plugin interface allows you to take over the packet send and receive functions, allowing you to pack the buffers as required by the protocol. The core read and write functions can (and should) be used by the protocol plugin.
.SS HTTP
.sp
This protocol plugin allows you to map HTTP requests to Gearman jobs. It only provides client job submission currently, but it may be extended to support other request types in the future. The plugin can handle both GET and POST data, the latter being used to send a workload to the job server. The URL being requested is translated into the function being called.
.sp
For example, the request:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
POST /reverse HTTP/1.1
Content\-Length: 12

Hello world!
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Is translated into a job submission request for the function “reverse” and workload “Hello world!”. This will respond with:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
HTTP/1.0 200 OK
X\-Gearman\-Job\-Handle: H:lap:4
Content\-Length: 12
Server: Gearman/0.8

!dlrow olleH
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following headers can be passed to change the behavior of the job:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
* X\-Gearman\-Unique: <unique key>
* X\-Gearman\-Background: true
* X\-Gearman\-Priority: <high|low>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For example, to run a low priority background job, the following request can be sent:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
POST /reverse HTTP/1.1
Content\-Length: 12
X\-Gearman\-Background: true
X\-Gearman\-Priority: low

Hello world!
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The response for this request will not have any data associated with it since it was a background job:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
HTTP/1.0 200 OK
X\-Gearman\-Job\-Handle: H:lap:6
Content\-Length: 0
Server: Gearman/0.8
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The HTTP protocol should be considered experimental.
.SH HOME
.sp
To find out more information please check:
\fI\%http://gearman.info/\fP
.SH SEE ALSO
.sp
\fIgearman(1)\fP \fIgearadmin(1)\fP \fIlibgearmand(3)\fP
.SH AUTHOR
Data Differential http://www.datadifferential.com/
.SH COPYRIGHT
2011-2013, Data Differential, http://www.datadifferential.com/
.\" Generated by docutils manpage writer.
.


flexviews/CHANGES
== 1.8.1 beta 2 ==
Views with aggregate functions lacked unique keys on their GROUP BY attributes.  This resulted in wrong results during refresh.  This has been corrected.
There is now a limit of 16 GROUP BY columns in a materialized view.  This is because MySQL will not allow more than 16 columns in an index.
A future version of Flexviews may lift this limitation using a hash column.

Views without aggregate tables now include a hash column.  This makes updating the views efficient without requiring the user explicitly create an 
index over columns in the table to facilitate fast refresh.

You need to disable all views without aggregate functions before upgrading and then re-enable them after upgrading.

== 1.8.0 beta 1 ==
The refresh algorithm for views without aggregate functions has been changed to take into account the order of changes in the database. 
All table changelogs now include a fv$gsn column to support the improved refresh algorithm
FlexCDC is improved to support all datatypes

== 1.6.1 GA ==
Whats new in Flexviews 1.6.1 GA
    * This is the first Flexviews GA release!
    * See UPGRADE if you are running an 1.6.0 RC1 or RC2.  Upgrade from earlier versions is not supported.
    * Flexviews now automatically adds indexes to changelogs to improve performance
    * Fixed a regression in flexviews.add_table() which would allow a table without a changelog to be added to an INCREMENTAL view 

== 1.6.0RC2 ==
Whats new in Flexviews 1.6.0RC2
    * This release is a performance and bug fix release, there are no new features.

Significant bug fixes
    * Reduce the number of subqueries and temporary tables
    * Eliminate full table scans during the delta application process
    * Remove dead code
    * Add indexes to GROUP and COUNT expressions automatically
    * When the first table (the one with a NULL join condition) was removed and then added again, the JOIN clause could have be constructed in an incorrect order.
    * Fix a problem with the PERCENTILE function
    * Improve some error messages

== 1.6.0RC1 ==
Whats new in Flexviews 1.6.0RC1

    * This is the first release candidate before the final release.  If no major bugs are uncovered, then the next release will be the first GA release. 
    * Flexviews now has a test suite for all major features.  The creation of these tests uncovered a number of issues which have been resolved in this release. 
    * All MySQL aggregate functions except GROUP_CONCAT are now supported. 
    * A special aggregate function called PERCENTILE is now also supported.  The calculation uses a modified version of the GROUP_CONCAT based solution suggested by Roland Bouman for percentiles.  This function should be considered experimental.  Please report bugs if you find any.
    * You can add indexes to enabled materialized views using SQL_API/add_expr
    * Adding PRIMARY KEY indexes is no longer supported.  All views get an auto_incrementing primary  key.  You can add additional UNIQUE indexes instead.
    * There is an upgrade process from 1.5.3b (see UPGRADE and upgrade.sql)

Significant bug fixes

    * Views with aggregate functions but no GROUP BY columns now work properly for all supported aggregate function types
    * NULL values in GROUP BY columns are now properly supported
    * NULL values now work properly with distributive aggregate functions
    * There is a wrapper script around run_consumer.php which can restart the consumer if it stops running

== 1.5.3b3 ==
Improvements for Flexviews, materialized view for MySQL 5.1+
--------------------------------------------------------------------
   1. There is up-to-date documentation available as manual.html, a robodoc generated version of the manual for the SQL_API interface. The manual is generated from RoboDOC comments embedded in the source code. This makes it easy to maintain the documentation if the interface changes.
   2. Flexviews has a new easier to read, and hopefully easier to understand website. You can find a copy of the manual on the web site as well.
   3. There is a SQL converter on the website than will automatically generate Flexviews SQL_API calls from a SELECT statement. The next step will be to turn this into a MySQL proxy script that can handle 'CREATE ALGORITHM=INCREMENTAL MATERIALIZED VIEW ...'
   4. Support for the FlexCDC auto_changelog mode, which automatically logs changes of all tables. This is great for small schemas.

Improvements for FlexCDC
---------------------------------
   1. The main wrapper scripts around flexcdc.php accidently included 'consumer.php' instead of 'flexcdc.php', which I had renamed shortly before the prior release. This problem has been resolved.
   2. I added an ini parameter to control the issuing of warnings by the script
   3. I added info to the README about when to not use auto_changelog
   4. flexcdc.php now looks for the ini file in the same directory in which the script resides, instead of the working directory
   5. the next release will have a command line option to specify the ini location. this will make it easy to run more than one consumer.



warpsql-plugins/extern/gearmand/libtest/gearmand.cc
/*  vim:expandtab:shiftwidth=2:tabstop=2:smarttab:
 *
 *  Data Differential YATL (i.e. libtest)  library
 *
 *  Copyright (C) 2012 Data Differential, http://datadifferential.com/
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are
 *  met:
 *
 *      * Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *      * Redistributions in binary form must reproduce the above
 *  copyright notice, this list of conditions and the following disclaimer
 *  in the documentation and/or other materials provided with the
 *  distribution.
 *
 *      * The names of its contributors may not be used to endorse or
 *  promote products derived from this software without specific prior
 *  written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "libtest/yatlcon.h"
#include <libtest/common.h>

#include <libtest/gearmand.h>

#include "libgearman/ssl.h"

using namespace libtest;

#include <cassert>
#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <signal.h>
#include <sstream>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#ifndef __INTEL_COMPILER
#pragma GCC diagnostic ignored "-Wold-style-cast"
#endif

using namespace libtest;

class Gearmand : public libtest::Server
{
private:
public:
  Gearmand(const std::string& host_arg, in_port_t port_arg, bool libtool_, const char* binary);

  bool ping()
  {
    reset_error();

    if (out_of_ban_killed())
    {
      return false;
    }

    SimpleClient client(_hostname, _port);

    std::string response;
    bool ret= client.send_message("version", response);

    if (ret == false)
    {
      ASSERT_TRUE_(client.is_error(), "client.send_message() failed but no error was set");
    }

    if (client.is_error())
    {
      error(client.error_file(), client.error_line(), client.error());
    }

    return ret;
  }

  const char *name()
  {
    return "gearmand";
  };

  void log_file_option(Application& app, const std::string& arg)
  {
    if (arg.empty() == false)
    {
      std::string buffer("--log-file=");
      buffer+= arg;
      // @note leave the logic as a placeholder
#if defined(VCS_CHECKOUT) && VCS_CHECKOUT
      app.add_option("--verbose=INFO");
#else
      app.add_option("--verbose=INFO");
#endif
      app.add_option(buffer);
    }
  }

  bool has_log_file_option() const
  {
    return true;
  }

  bool is_libtool()
  {
    return true;
  }

  bool has_syslog() const
  {
    return false; //  --syslog.errmsg-enable
  }

  bool has_port_option() const
  {
    return true;
  }

  bool build();
};

Gearmand::Gearmand(const std::string& host_arg, in_port_t port_arg, bool libtool_, const char* binary_arg) :
  libtest::Server(host_arg, port_arg, binary_arg, libtool_)
{
  set_pid_file();
}

bool Gearmand::build()
{
  if (getuid() == 0 or geteuid() == 0)
  {
    add_option("-u", "root");
  }

  add_option("--listen=localhost");


  if (is_ssl())
  {
#if defined(HAVE_SSL) && HAVE_SSL
    add_option("--ssl");
    add_option("--ssl-ca-file=" YATL_CA_CERT_PEM);
    add_option("--ssl-certificate=" YATL_CERT_PEM);
    add_option("--ssl-key=" YATL_CERT_KEY_PEM);
#endif
  }

  return true;
}

namespace libtest {

libtest::Server *build_gearmand(const char *hostname, in_port_t try_port, const char* binary)
{
  if (binary == NULL)
  {
#if defined(HAVE_GEARMAND_BINARY)
# if defined(GEARMAND_BINARY)
    if (HAVE_GEARMAND_BINARY)
    {
      binary= GEARMAND_BINARY;
    }
# endif
#endif
  }

  if (binary == NULL)
  {
    return NULL;
  }

  bool is_libtool_script= true;

  if (binary[0] == '/')
  {
    is_libtool_script= false;
  }

  return new Gearmand(hostname, try_port, is_libtool_script, binary);
}

}
